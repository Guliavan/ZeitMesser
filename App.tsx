import React, {Component} from 'react';
import { Svg, G, Circle, Path } from 'react-native-svg';
import { StyleSheet, Text, View, TextInput } from 'react-native';

const DEFAULT_TIME = 90;
const FULL_DASH_ARRAY = 283; //(2 * pi * 45 (r))
export const WARNING_THRESHOLD = 10;
export const ALERT_THRESHOLD = 5;
export const COLOR_CODES = {
    info: {
        color: "#41b883"
    },
    warning: {
        color: "#ffa500",
        threshold: WARNING_THRESHOLD
    },
    alert: {
        color: "#ff0000",
        threshold: ALERT_THRESHOLD
    }
};
const { alert, warning, info } = COLOR_CODES;

const styles = StyleSheet.create({
    container: {
        position: 'relative',
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        justifyContent: 'center',
    },
    timerContainer: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    baseTimer: {
        position: 'relative'
    },
    baseTimerSvg: {
        transform: [
           { scale: -3 },
           { scaleX: -1 },
           { rotate: '-90deg' }
        ]
    },
    baseTimerPathRemaining: {
        transform: [
            { rotate: '90deg' }
        ]
    },
    baseTimerCountdownContainer: {
        position: 'absolute'
    },
    baseTimerCountdownText: {
        fontSize: 100,
        color: '#333'
    },
    baseTimerInputContainer: {
        position: 'relative',
        top: 140
    },
    baseTimerInput: {
        height: 120,
        borderWidth: 0,
        fontSize: 50,
        color: 'white',
        padding: 10,
        width: 500,
        textAlign: 'center',
        position: 'absolute'
    },
    baseTimerInputText: {
        marginTop: 10,
        fontSize: 40,
    }
});

export default class App extends Component {
    constructor(props:Object) {
        super(props);
        this.state = {
            startTime: DEFAULT_TIME,
            timeLeft: DEFAULT_TIME,
            intervalTimer: 0,
            pathColor: COLOR_CODES.info.color,
            strokeDasharray: '283 283'
        };
        this.startTimer = this.startTimer.bind(this);
        this.resetTimer = this.resetTimer.bind(this);
        this.iterateInterval = this.iterateInterval.bind(this);
        this.setCircleDasharray = this.setCircleDasharray.bind(this);
        this.calculateTimeFraction = this.calculateTimeFraction.bind(this);
        this.setPathColor = this.setPathColor.bind(this);
        this.setRemainingPathColor = this.setRemainingPathColor.bind(this);
        this.onChangeStartTime = this.onChangeStartTime.bind(this);
    }

    resetTimer () {
        this.setState({timeLeft: this.state.startTime});
        if (this.state.intervalTimer != 0) {
            this.setPathColor(info.color);
            clearInterval(this.state.intervalTimer);
        }
    };

    iterateInterval () {
        if (this.state.timeLeft <= 0) {
            this.resetTimer();
        } else {
            this.setState({timeLeft: this.state.timeLeft - 1});
            this.setCircleDasharray();
            this.setRemainingPathColor();
        }
    };

    startTimer() {
        this.resetTimer();
        const newTimer = setInterval( this.iterateInterval, 1000 );
        this.setState({ intervalTimer: newTimer });
    };

    formatTimeLeft(time: number) {
        const minutes:number = Math.floor(time / 60);
        let seconds = time % 60;
        let minutesPrint:string = `${minutes}`;
        if (minutes < 1) {
            minutesPrint = `0`;
        }
        let secondsPrint:string = `${seconds}`;
        if (seconds < 10) {
            secondsPrint = `0${seconds}`;
        }
        return `${minutesPrint}:${secondsPrint}`;
    };

    calculateTimeFraction() {
        const rawTimeFraction:Number = this.state.timeLeft / this.state.startTime;
        return rawTimeFraction - (1 / this.state.startTime) * (1 - rawTimeFraction);
    };

    setCircleDasharray() {
        const circleDasharray = `${(
            this.calculateTimeFraction() * FULL_DASH_ARRAY
        ).toFixed(0)} 283`;
        this.setState({ strokeDasharray: circleDasharray });
    };

    setPathColor(color:string) {
        this.setState({ pathColor: color });
    };

    setRemainingPathColor() {
        if (this.state.timeLeft <= alert.threshold) {
            this.setPathColor(alert.color);
        } else if (this.state.timeLeft <= warning.threshold) {
            this.setPathColor(warning.color);
        }
    };

    onChangeStartTime(value:string) {
        let numericalValue:number = 0;
        const reverseString = (text:string) => text.split('').reverse().join('');
        if (value.length > 2) {
            const reversedValue = reverseString(value);
            const seconds = parseInt(reverseString(reversedValue.substr(0, 2)));
            const minutes = parseInt(reverseString(reversedValue.substr(2))) * 60;
            numericalValue = minutes + seconds
        } else if (value.length > 0) {
            numericalValue = parseInt(value);
        }
        this.setState({ startTime: numericalValue, timeLeft: numericalValue });
    };

    componentWillUnmount() {
        if (this.state.intervalTimer != 0) {
            clearInterval(this.state.intervalTimer);
        }
    }

    render() {
        return (
            <View style={styles.container}>
                <View style={styles.timerContainer}>
                    {/* <Text>Open up App.tsx to start working on your app!</Text> */}
                    <View style={styles.baseTimer}>
                        <Svg
                            testID="baseTimerPathSVG"
                            style={styles.baseTimerSvg}
                            height="100px"
                            width="100px"
                            onPress={this.startTimer}>
                            <G>
                                <Circle cx="50" cy="50" r="45" fill="white"/>
                                <Path
                                    d="
                                    M 50, 50
                                    m -45, 0
                                    a 45,45 0 1,0 90,0
                                    a 45,45 0 1,0 -90,0
                                    "
                                    fill="white"
                                    strokeLinecap="round"
                                    strokeWidth="7"
                                    strokeDasharray={this.state.strokeDasharray}
                                    stroke={this.state.pathColor}></Path>
                            </G>
                        </Svg>
                    </View>
                    <View style={styles.baseTimerCountdownContainer}>
                        <Text style={styles.baseTimerCountdownText} onPress={this.startTimer}>
                            {this.formatTimeLeft(this.state.timeLeft)}
                        </Text>
                    </View>
                </View>
                <View style={styles.baseTimerInputContainer}>
                    <Text style={styles.baseTimerInputText}>Change time</Text>
                    <TextInput
                        testID="timerTextInput"
                        style={styles.baseTimerInput}
                        onChangeText={value => this.onChangeStartTime(value)}
                        value={`${this.state.startTime}`}
                        maxLength={3}
                        keyboardType='number-pad'
                        enablesReturnKeyAutomatically={true}
                        returnKeyType='done'
                        caretHidden={true}
                        />
                </View>
            </View>
        );
    }
}


