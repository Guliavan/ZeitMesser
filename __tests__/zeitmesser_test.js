import React from 'react';
import App, {COLOR_CODES} from '../App.tsx';

import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-native-testing-library';

it('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
});

jest.useFakeTimers();

const DEFAULT_TIME = '1:30';

it ('starts the timer upon user click timer text', () => {
    const { getByText } = render(<App />);

    const timerText = getByText(DEFAULT_TIME);
    fireEvent.press(timerText);
    jest.advanceTimersByTime(60000);
    expect(timerText.props.children).toEqual('0:30');
});

it ('changes the timer text upon input change', () => {
    const { getByText, getByTestId } = render(<App />);

    const timerText = getByText(DEFAULT_TIME);
    const timerTextInput = getByTestId('timerTextInput');
    // Test seconds change
    fireEvent.changeText(timerTextInput, '20');
    expect(timerText.props.children).toEqual('0:20');
    // Test minute/seconds change with numbers < 6
    fireEvent.changeText(timerTextInput, '120');
    expect(timerText.props.children).toEqual('1:20');
    // Test minute/seconds change with numbers > 6
    fireEvent.changeText(timerTextInput, '160');
    expect(timerText.props.children).toEqual('2:00');
});

it ('has startup color on normal timer value (> 10 sec)', () => {
    const { getByText, getByTestId } = render(<App />);

    const timerText = getByText(DEFAULT_TIME);
    const timerSVG = getByTestId('baseTimerPathSVG');
    const timerSVGProps = timerSVG.props.children.props.children.props.children[1].props;

    // Start timer
    expect(timerSVGProps.stroke).toEqual(COLOR_CODES.info.color);
    expect(timerText.props.children).toEqual('1:30');
    fireEvent.press(timerText);

    jest.advanceTimersByTime(79000);
    expect(timerText.props.children).toEqual('0:11');
    const timerSVGPropsAfter = timerSVG.props.children.props.children.props.children[1].props;
    expect(timerSVGPropsAfter.stroke).toEqual(COLOR_CODES.info.color);
});

it ('changes the color on warn timer value (<= 10 sec)', () => {
    const { getByText, getByTestId } = render(<App />);

    const timerText = getByText(DEFAULT_TIME);
    const timerSVG = getByTestId('baseTimerPathSVG');
    const timerSVGProps = timerSVG.props.children.props.children.props.children[1].props;

    // Start timer
    expect(timerSVGProps.stroke).toEqual(COLOR_CODES.info.color);
    fireEvent.press(timerText);

    jest.advanceTimersByTime(80000);
    expect(timerText.props.children).toEqual('0:10');
    const timerSVGPropsAfter = timerSVG.props.children.props.children.props.children[1].props;
    expect(timerSVGPropsAfter.stroke).toEqual(COLOR_CODES.warning.color);
});

it ('changes the color on alert timer value (<= 5 sec)', () => {
    const { getByText, getByTestId } = render(<App />);

    const timerText = getByText(DEFAULT_TIME);
    const timerSVG = getByTestId('baseTimerPathSVG');
    const timerSVGProps = timerSVG.props.children.props.children.props.children[1].props;

    // Start timer
    expect(timerSVGProps.stroke).toEqual(COLOR_CODES.info.color);
    fireEvent.press(timerText);

    jest.advanceTimersByTime(85000);
    expect(timerText.props.children).toEqual('0:05');
    const timerSVGPropsAfter = timerSVG.props.children.props.children.props.children[1].props;
    expect(timerSVGPropsAfter.stroke).toEqual(COLOR_CODES.alert.color);
});