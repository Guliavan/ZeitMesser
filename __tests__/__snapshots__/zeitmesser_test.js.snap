// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <RNSVGSvgView
        bbHeight="100px"
        bbWidth="100px"
        focusable={false}
        height="100px"
        onPress={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        responsible={true}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "transform": Array [
                Object {
                  "scale": -3,
                },
                Object {
                  "scaleX": -1,
                },
                Object {
                  "rotate": "-90deg",
                },
              ],
            },
            Object {
              "opacity": 1,
            },
            Object {
              "flex": 0,
              "height": 100,
              "width": 100,
            },
          ]
        }
        testID="baseTimerPathSVG"
        width="100px"
      >
        <RNSVGGroup
          fill={4278190080}
          fillOpacity={1}
          fillRule={1}
          font={Object {}}
          matrix={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          opacity={1}
          originX={0}
          originY={0}
          propList={Array []}
          rotation={0}
          scaleX={1}
          scaleY={1}
          skewX={0}
          skewY={0}
          stroke={null}
          strokeDasharray={null}
          strokeDashoffset={null}
          strokeLinecap={0}
          strokeLinejoin={0}
          strokeMiterlimit={4}
          strokeOpacity={1}
          strokeWidth={1}
          vectorEffect={0}
          x={0}
          y={0}
        >
          <RNSVGGroup
            fill={4278190080}
            fillOpacity={1}
            fillRule={1}
            font={Object {}}
            matrix={
              Array [
                1,
                0,
                0,
                1,
                0,
                0,
              ]
            }
            opacity={1}
            originX={0}
            originY={0}
            propList={Array []}
            rotation={0}
            scaleX={1}
            scaleY={1}
            skewX={0}
            skewY={0}
            stroke={null}
            strokeDasharray={null}
            strokeDashoffset={null}
            strokeLinecap={0}
            strokeLinejoin={0}
            strokeMiterlimit={4}
            strokeOpacity={1}
            strokeWidth={1}
            vectorEffect={0}
            x={0}
            y={0}
          >
            <RNSVGCircle
              cx="50"
              cy="50"
              fill={4294967295}
              fillOpacity={1}
              fillRule={1}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
              opacity={1}
              originX={0}
              originY={0}
              propList={
                Array [
                  "fill",
                ]
              }
              r="45"
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={null}
              strokeDasharray={null}
              strokeDashoffset={null}
              strokeLinecap={0}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth={1}
              vectorEffect={0}
              x={0}
              y={0}
            />
            <RNSVGPath
              d=" M 50, 50 m -45, 0 a 45,45 0 1,0 90,0 a 45,45 0 1,0 -90,0 "
              fill={4294967295}
              fillOpacity={1}
              fillRule={1}
              matrix={
                Array [
                  1,
                  0,
                  0,
                  1,
                  0,
                  0,
                ]
              }
              opacity={1}
              originX={0}
              originY={0}
              propList={
                Array [
                  "fill",
                  "stroke",
                  "strokeWidth",
                  "strokeDasharray",
                  "strokeLinecap",
                ]
              }
              rotation={0}
              scaleX={1}
              scaleY={1}
              skewX={0}
              skewY={0}
              stroke={4282497155}
              strokeDasharray={
                Array [
                  "283",
                  "283",
                ]
              }
              strokeDashoffset={null}
              strokeLinecap={1}
              strokeLinejoin={0}
              strokeMiterlimit={4}
              strokeOpacity={1}
              strokeWidth="7"
              vectorEffect={0}
              x={0}
              y={0}
            />
          </RNSVGGroup>
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <View
      style={
        Object {
          "position": "absolute",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Object {
            "color": "#333",
            "fontSize": 100,
          }
        }
      >
        1:30
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "position": "relative",
        "top": 140,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 40,
          "marginTop": 10,
        }
      }
    >
      Change time
    </Text>
    <TextInput
      allowFontScaling={true}
      caretHidden={true}
      enablesReturnKeyAutomatically={true}
      keyboardType="number-pad"
      maxLength={3}
      onChangeText={[Function]}
      rejectResponderTermination={true}
      returnKeyType="done"
      style={
        Object {
          "borderWidth": 0,
          "color": "white",
          "fontSize": 50,
          "height": 120,
          "padding": 10,
          "position": "absolute",
          "textAlign": "center",
          "width": 500,
        }
      }
      testID="timerTextInput"
      underlineColorAndroid="transparent"
      value="90"
    />
  </View>
</View>
`;
